### PR TITLE
LTTP: Move more options to options system

### PR DESCRIPTION
**Pull request by alwaysintreble**
This moves over a bunch of options:
- The remaining logic options. `logic` is now `glitches_required`
- Enemy options
- Prize shuffle
- Triforce options
- Medallion options (with a fix for random default value if not defined in the YAML)
- Dungeon counter mode.
- `mode` renamed to `world_state` to be more clear and moved over
- Item pool and functionality
- Sprite pool is now an optionlist

Framework for other sprite options is present but not finished. Actual sprite is ready and will be a textchoice when #728 is merged. RandomSprite I still haven't decided the best course as far as naming convention goes. As far as implementation planning on this being split into a togle and option list as all of the options can be enabled at the same time.

### CODE REVIEW

**Comment by alwaysintreble** [comment]
Should address all of the missing docstrings from #823

File: BaseClasses.py [hunk 1]
```diff
@@ -88,7 +87,7 @@ def __init__(self, players: int):
         self.shuffle_ganon = True
         self.spoiler = Spoiler(self)
         self.fix_trock_doors = self.AttributeProxy(
-            lambda player: self.shuffle[player] != 'vanilla' or self.mode[player] == 'inverted')
+            lambda player: self.shuffle[player] != 'vanilla' or self.world_state[player] == 2)
```
Comment (black-sliver) [inline BaseClasses.py:90]: I don't like using ints here, same for item_functionality. Shouldn't that be a regular Options.Choice?
I'm guessing the problem is that this is still in core?

File: BaseClasses.py [hunk 2]
```diff
@@ -1370,7 +1352,7 @@ def write_option(option_key: str, option_obj: type(Options.Option)):
             res = getattr(self.world, option_key)[player]
             display_name = getattr(option_obj, "display_name", option_key)
             try:
-                outfile.write(f'{display_name + ":":33}{res.get_current_option_name()}\n')
+                outfile.write(f'{display_name + ":":33}{res.get_current_option_name}\n')
```
Comment (black-sliver) [inline BaseClasses.py:1355]: if you change it to a property, please drop the `get_`.

File: Generate.py [hunk 3]
```diff
@@ -309,8 +309,10 @@ def get_choice_legacy(option, root, value=None) -> Any:
 def get_choice(option, root, value=None) -> Any:
     if option not in root:
         return value
-    if type(root[option]) is list:
-        return random.choices(root[option])[0]
+    if isinstance(root[option], list):
+        if len(root[option]) > 0:
+            return random.choices(root[option])[0]
+        return value
```
Comment (black-sliver) [inline Generate.py:315]: why/where is this required?

File: Generate.py [hunk 4]
```diff
@@ -514,7 +516,11 @@ def handle_option(ret: argparse.Namespace, game_weights: dict, option_key: str,
             if hasattr(player_option, "verify"):
                 player_option.verify(AutoWorldRegister.world_types[ret.game])
     else:
-        setattr(ret, option_key, option(option.default))
+        if option.default == "random":
```
Comment (black-sliver) [inline Generate.py:519]: this can be changed to `elif`. Wondering why this change is required though.

File: Options.py [hunk 5]
```diff
@@ -82,7 +82,7 @@ class Option(typing.Generic[T], metaclass=AssembleOptions):
     options: typing.Dict[str, int]
 
     def __repr__(self) -> str:
-        return f"{self.__class__.__name__}({self.get_current_option_name()})"
+        return f"{self.__class__.__name__}({self.get_current_option_name})"
```
Comment (black-sliver) [inline Options.py:85]: if you change it to a property, please drop the `get_`

File: worlds/alttp/ItemPool.py [hunk 6]
```diff
@@ -226,17 +226,11 @@ def generate_itempool(world):
     player = world.player
     world = world.world
 
-    if world.difficulty[player] not in difficulties:
-        raise NotImplementedError(f"Diffulty {world.difficulty[player]}")
     if world.goal[player] not in {'ganon', 'pedestal', 'bosses', 'triforcehunt', 'localtriforcehunt', 'icerodhunt',
                                   'ganontriforcehunt', 'localganontriforcehunt', 'crystals', 'ganonpedestal'}:
         raise NotImplementedError(f"Goal {world.goal[player]} for player {player}")
```
Comment (black-sliver) [inline worlds/alttp/ItemPool.py:231]: ~~you forgot to fix goal~~

File: worlds/alttp/Options.py [hunk 7]
```diff
@@ -1,31 +1,33 @@
...
+    """Determine the logic used to reach locations."""
+    display_name = "Glitch Logic"
     option_no_glitches = 0
     option_minor_glitches = 1
     option_overworld_glitches = 2
     option_hybrid_major_glitches = 3
     option_no_logic = 4
+    alias_none = option_no_glitches
     alias_owg = 2
     alias_hmg = 3
```
Comment (black-sliver) [inline worlds/alttp/Options.py:17]: this breaks me. please make them look identical (either all int or all option_)

File: worlds/alttp/Options.py [hunk 8]
```diff
@@ -79,41 +80,44 @@ def hints_useful(self):
...
 
 
 class map_shuffle(DungeonItem):
-    """Map Placement"""
+    """How maps will be placed."""
     item_name_group = "Maps"
     display_name = "Map Shuffle"
 
 
 class Crystals(Range):
```
Comment (black-sliver) [inline worlds/alttp/Options.py:107]: wondering if this should be named CrystalsRange to sound more generic

File: worlds/alttp/Options.py [hunk 9]
```diff
@@ -122,22 +126,151 @@ class TriforcePieces(Range):
...
+
+
+class TriforceExtra(TriforcePieces):
+    """How many extra Triforce pieces are available."""
+    display_name = "Extra Triforce Pieces"
+    default = 10
+
+
+class TriforcePercentage(Range):
+    """Percentage of required Triforce pieces that will be available."""
```
Comment (black-sliver) [inline worlds/alttp/Options.py:148]: probably for a separate PR, but at some point we should have a `display_if = {"goal": ("trifoce hunt",)}` or something that hides the option in the UI if the requirement is not statisfied.

File: worlds/alttp/Options.py [hunk 10]
```diff
@@ -318,20 +529,126 @@ class MenuSpeed(Choice):
...
+                  "Super Meat Boy", "Susie", "Swatchy", "Swiper", "Tanooki Mario", "TASBot", "Tea Time", "Terra (Esper)",
+                  "Terry", "Tetra Sheet", "TGH", "The Professor", "The Pug", "Thief", "ThinkDorm", "Thomcrow", "Tile",
+                  "Tingle", "TMNT", "Toad", "Toadette", "TotemLinks", "TP Zelda", "Trogdor the Burninator",
+                  "Tunic Color Link", "TwoFaced", "Ty the Tasmanian Tiger", "Ultros", "Umbreon", "Valeera",
+                  "VanillaLink", "Vaporeon", "Vegeta", "Vera", "Vitreous", "Vivi", "Vivian", "Wario", "White Mage",
+                  "Will", "Wizzrobe", "Wolf Link (Festive)", "Wolf Link (TP)", "Yoshi", "Yunica Tovah", "Zandra",
+                  "Zaruvyen", "Zebra Unicorn", "Zeckemyro", "Zelda", "Zero Suit Samus", "Zora"}
+
+
+# # TODO obviously waiting on the freetext options
```
Comment (black-sliver) [inline worlds/alttp/Options.py:611]: :cry:

File: worlds/alttp/Rom.py [hunk 11]
```diff
@@ -37,7 +38,7 @@
 from Utils import local_path, user_path, int16_as_bytes, int32_as_bytes, snes_to_pc, is_frozen
 from worlds.alttp.Items import ItemFactory, item_table, item_name_groups, progression_items
 from worlds.alttp.EntranceShuffle import door_addresses
-from worlds.alttp.Options import smallkey_shuffle
+from worlds.alttp.Options import smallkey_shuffle, PrizeShuffle
```
Comment (black-sliver) [inline worlds/alttp/Rom.py:41]: smallkey_shuffle should be renamed to SmallkeyShuffle, but i guess that can be done separately

File: worlds/alttp/Rom.py [hunk 12]
```diff
@@ -1440,14 +1444,14 @@ def chunk(l, n):
...
+    rom.write_byte(0x18004A, 0x00 if world.world_state[player] != WorldState.option_inverted else 0x01)  # Inverted mode
     rom.write_byte(0x18005D, 0x00)  # Hammer always breaks barrier
-    rom.write_byte(0x2AF79, 0xD0 if world.mode[
-                                        player] != 'inverted' else 0xF0)  # vortexes: Normal  (D0=light to dark, F0=dark to light, 42 = both)
-    rom.write_byte(0x3A943, 0xD0 if world.mode[
-                                        player] != 'inverted' else 0xF0)  # Mirror: Normal  (D0=Dark to Light, F0=light to dark, 42 = both)
-    rom.write_byte(0x3A96D, 0xF0 if world.mode[
-                                        player] != 'inverted' else 0xD0)  # Residual Portal: Normal  (F0= Light Side, D0=Dark Side, 42 = both (Darth Vader))
+    rom.write_byte(0x2AF79, 0xD0 if world.world_state[
+                                        player] != WorldState.option_inverted else 0xF0)  # vortexes: Normal  (D0=light to dark, F0=dark to light, 42 = both)
```
Comment (black-sliver) [inline worlds/alttp/Rom.py:1450]: should probably have a local variable `inverted = ...` to make this more readable
that's used a ton of times above and below as well.

File: worlds/alttp/Shops.py [hunk 13]
```diff
@@ -301,7 +301,7 @@ def create_shops(world, player: int):
                         new_items = new_dark_shop
                 keeper = world.random.choice([0xA0, 0xC1, 0xFF])
                 player_shop_table[name] = ShopData(typ, shop_id, keeper, custom, locked, new_items, sram_offset)
-    if world.mode[player] == "inverted":
+    if world.world_state[player] == 2:
```
Comment (black-sliver) [inline worlds/alttp/Shops.py:304]: If the int is because importing Options breaks up top, you can just import right below `def create_shops`

File: worlds/alttp/__init__.py [hunk 14]
```diff
@@ -427,12 +442,12 @@ def stage_fill_hook(cls, world, progitempool, nonexcludeditempool, localrestitem
         trash_counts = {}
         standard_keyshuffle_players = set()
         for player in world.get_game_players("A Link to the Past"):
-            if world.mode[player] == 'standard' and world.smallkey_shuffle[player] \
+            if world.world_state[player] == 1 and world.smallkey_shuffle[player] \
```
Comment (black-sliver) [inline worlds/alttp/__init__.py:445]: why the int?

**Review by black-sliver — CHANGES_REQUESTED** [review]
Found a few things.

I have not checked most of the changes in `__init__.py` and `Rules.py`, because some program flow changed and i don't know what glitch should do what.

File: Generate.py (hunk 4 above)
Comment (alwaysintreble) [reply Generate.py:519]: There was recently a PR to allow random as the default value. If random is the default value for an option and said option is not present in a submitted yaml file it will assign "random" as the value instead of making a random choice. By calling the options `from_any` what happens can still be controlled by the option.

File: worlds/alttp/ItemPool.py (hunk 6 above)
Comment (alwaysintreble) [reply worlds/alttp/ItemPool.py:231]: :)

File: worlds/alttp/Options.py (hunk 8 above)
Comment (alwaysintreble) [reply worlds/alttp/Options.py:107]: i mean it could be? but it's only used for 2 options here so eh

File: worlds/alttp/Options.py (hunk 9 above)
Comment (alwaysintreble) [reply worlds/alttp/Options.py:148]: i think the best way would be having some sort of `option_requires` where you can define an option and a specified value for said option. This would allow for webhost options to dynamically gray themselves out etc too.

File: worlds/alttp/Rom.py (hunk 12 above)
Comment (alwaysintreble) [reply worlds/alttp/Rom.py]: writing a property when you're almost done needing it and then not going back to use it? couldn't be me.

**Comment by black-sliver** [comment]
```
  File "/worlds/alttp/Rom.py", line 1852, in apply_rom_settings
    {'normal': 0x00, 'hide_goal': 0x01, 'hide_required': 0x02, 'hide_both': 0x03}[triforcehud]
KeyError: 'always_show'
```

File: BaseClasses.py [hunk 15]
```diff
@@ -829,7 +808,7 @@ def is_not_bunny(self, region: Region, player: int) -> bool:
         if self.has('Moon Pearl', player):
             return True
 
-        return region.is_light_world if self.world.mode[player] != 'inverted' else region.is_dark_world
+        return region.is_light_world if self.world.world_state[player] != 2 else region.is_dark_world
```
Comment (espeon65536) [inline BaseClasses.py:811]: The usage of `WorldState.inverted` is inconsistent here and in many other places. They should all be changed to use the property rather than direct comparison.

File: worlds/alttp/Bosses.py [hunk 16]
```diff
@@ -123,7 +123,7 @@ def GanonDefeatRule(state, player: int) -> bool:
     can_hurt = state.has_beam_sword(player)
     common = can_hurt and state.has_fire_source(player)
     # silverless ganon may be needed in anything higher than no glitches
-    if state.world.logic[player] != 'noglitches':
+    if state.world.glitches_required[player] != 0:
```
Comment (espeon65536) [inline worlds/alttp/Bosses.py:126]: Might want to create `glitchless` and `glitched` properties like the `inverted` property, although this gets used less often so it's not as important for readability.

File: worlds/alttp/EntranceShuffle.py [hunk 17]
```diff
@@ -1948,17 +1951,18 @@ def connect_random(world, exitlist, targetlist, player, two_way=False):
 
 
 def connect_mandatory_exits(world, entrances, caves, must_be_exits, player):
-
+    standard = world.world_state[player] == WorldState.option_standard
+    inverted = world.world_state[player] == WorldState.option_inverted
     # Keeps track of entrances that cannot be used to access each exit / cave
-    if world.mode[player] == 'inverted':
+    if standard:
```
Comment (espeon65536) [inline worlds/alttp/EntranceShuffle.py:1957]: This should be `inverted`

File: worlds/alttp/Rom.py [hunk 18]
```diff
@@ -1193,23 +1205,24 @@ def chunk(l, n):
...
         rom.write_bytes(0x180190, [0x02, 0x01, 0x00])  # set stopwatch mode
-    elif world.clock_mode[player] == 'countdown':
+    elif clock_mode == 'countdown':
         rom.write_bytes(0x180190, [0x01, 0x01, 0x00])  # set countdown, with no reset available
     else:
         rom.write_bytes(0x180190, [0x00, 0x00, 0x00])  # turn off clock mode
 
     # Set up requested clock settings
-    if world.clock_mode[player] in ['countdown-ohko', 'stopwatch', 'countdown']:
+    if world.worlds[player].clock_mode in ['countdown-ohko', 'stopwatch', 'countdown']:
```
Comment (espeon65536) [inline worlds/alttp/Rom.py:1225]: Use `clock_mode` from 1214 (and in later places as well)

File: worlds/alttp/Rules.py [hunk 19]
```diff
@@ -38,37 +45,31 @@ def set_rules(world):
...
-    elif world.mode[player] == 'standard':
+    elif world.world_state[player] == WorldState.option_standard:
         standard_rules(world, player)
-    elif world.mode[player] == 'inverted':
+    elif world.world_state[player].inverted:
         open_rules(world, player)
         inverted_rules(world, player)
     else:
-        raise NotImplementedError(f'World state {world.mode[player]} is not implemented yet')
+        raise NotImplementedError(f'World state {world.world_state[player]} is not implemented yet')
```
Comment (espeon65536) [inline worlds/alttp/Rules.py:59]: This exception doesn't need to exist since it's not even possible to get here without one of those three, so it should just be removed.

File: worlds/alttp/Rules.py [hunk 20]
```diff
@@ -215,7 +213,7 @@ def global_rules(world, player):
 
     set_rule(world.get_entrance('Sewers Door', player),
              lambda state: state._lttp_has_key('Small Key (Hyrule Castle)', player) or (
-                         world.smallkey_shuffle[player] == smallkey_shuffle.option_universal and world.mode[
+                         world.smallkey_shuffle[player] == smallkey_shuffle.option_universal and world.world_state[
                      player] == 'standard'))  # standard universal small keys cannot access the shop
```
Comment (espeon65536) [inline worlds/alttp/Rules.py:217]: This shouldn't be the string `'standard'` anymore

**Review by espeon65536 — CHANGES_REQUESTED** [review]
Only found a few minor things to change. However, it is unclear to me why the goal and entrance shuffle options aren't moved over, since those are two of the most important options. I know entrance shuffle has the seed/group issues, but I would like to at least see the goal option moved over.

File: BaseClasses.py (hunk 15 above)
Comment (alwaysintreble) [reply BaseClasses.py:811]: i didn't think to create the property until it was too late ok? i forgot some...

File: BaseClasses.py [hunk 21]
```diff
@@ -766,9 +745,9 @@ def can_extend_magic(self, player: int, smallmagic: int = 16,
         elif self.has('Magic Upgrade (1/2)', player):
             basemagic = 16
         if self.can_buy_unlimited('Green Potion', player) or self.can_buy_unlimited('Blue Potion', player):
-            if self.world.item_functionality[player] == 'hard' and not fullrefill:
+            if self.world.item_functionality[player] == 2 and not fullrefill:
```
Comment (Berserker66) [inline BaseClasses.py:748]: looking later at this I will have no idea which one is 2 or 3, please compare with the option key variable or name

File: Options.py [hunk 22]
```diff
@@ -427,7 +428,6 @@ def __init__(self, value: typing.Union[str, int]):
         assert isinstance(value, str) or isinstance(value, int), \
             f"{value} is not a valid option for {self.__class__.__name__}"
         self.value = value
-        super(TextChoice, self).__init__()
```
Comment (Berserker66) [inline Options.py:430]: This seems like a random change to be in this PR

File: worlds/alttp/EntranceShuffle.py [hunk 23]
```diff
@@ -1067,10 +1070,10 @@ def connect_reachable_exit(entrance, caves, doors):
         raise NotImplementedError(
             f'{world.shuffle[player]} Shuffling not supported yet. Player {world.get_player_name(player)}')
 
-    if world.logic[player] in ['owglitches', 'hybridglitches', 'nologic']:
+    if world.glitches_required[player] > 1:
```
Comment (Berserker66) [inline worlds/alttp/EntranceShuffle.py:1073]: > owglitches or the list-like approach would be preferred.

File: worlds/alttp/Options.py [hunk 24]
```diff
@@ -440,20 +672,119 @@ class MenuSpeed(Choice):
...
+class RandomSpriteToggle(Toggle):
+    """Enabling this will enable options in the random sprite event list, causing your sprite to be randomized from your
+    sprite pool."""
+    display_name = "Use Random Sprite Events"
+
+
+class SpritePool(OptionList):
+    """Limits the pool of available sprites for random events to specified pool."""
+    display_name = "Sprite Pool"
+    valid_keys = {"Abigail", "Adol", "Adventure 2600", "Aggretsuko", "Alice", "Angry Video Game Nerd", "Arcane",
```
Comment (Berserker66) [inline worlds/alttp/Options.py:690]: This should really not be hardcoded

File: worlds/alttp/Rom.py [hunk 25]
```diff
@@ -222,24 +222,33 @@ def apply_random_sprite_on_event(rom: LocalRom, sprite, local_random, allow_rand
             rom.write_byte(0x186381, 0x00 if userandomsprites else 0x01)
 
         onevent = 0
-        if sprite == 'randomonall':
+        if sprite == 7:
```
Comment (Berserker66) [inline worlds/alttp/Rom.py:225]: Ah yes, "sprite == 7", surely we will all know what that means in a few months

**Review by Berserker66 — CHANGES_REQUESTED** [review]

File: Options.py (hunk 22 above)
Comment (alwaysintreble) [reply Options.py:430]: this turned out to break the subclasses which i missed since i was overriding this in the Bosses class. I addressed this in a different PR and just mimicked it here to save me some time

**Comment by Berserker66** [comment]
It's been a while, admittedly, but we agreed that goal should be the last option moved over so that people realize that options are missing.

File: Options.py (hunk 22 above)
Comment (Berserker66) [reply Options.py:430]: which PR is that?

Comment (alwaysintreble) [reply Options.py:430]: #1044
